### PR TITLE
Phone number field type

### PR DESCRIPTION
**Pull request by marharyta**
- [x] Link to issue #2552
- [ ] Consider adding screenshots for ease of review

Screenshot 2021-03-11 at 13 46 22

## Documentation
- [ ] Update changelog if necessary
- [ ] API is documented and shows up in Swagger UI
- [ ] Components are added to guide page
- [ ] Update docs/ (if applicable)
- [ ] Update manual/ (if applicable)
- [ ] ADR for major architectural decisions or experiments
- [ ] New config options in config-defaults.edn

## Testing
- [ ] Complex logic is unit tested
- [ ] Valuable features are integration / browser / acceptance tested automatically

## Follow-up
- [X] New tasks are created for pending or remaining tasks - https://github.com/CSCfi/rems/projects/1#card-57027675

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -34,6 +34,7 @@ Changes since v2.16
 ### Additions
 - The example-theme now uses the Lato font.
 - The field id can now be changed in the form editor. (#1804)
+- Add phone nmber to fiels types in the form. (#2552)
```
Comment (opqdonut) [inline CHANGELOG.md:37]: typos: nmber, fiels

File: resources/translations/fi.edn [hunk 2]
```diff
@@ -266,6 +266,7 @@
                 :type-label "Lisätietokenttä"
                 :type-multiselect "Monivalintalista"
                 :type-option "Valintalista"
+                :type-phone-number "Phone number"
```
Comment (opqdonut) [inline resources/translations/fi.edn:269]: forgot to translate this?

File: resources/translations/sv.edn [hunk 3]
```diff
@@ -266,6 +266,7 @@
                 :type-label "Informationsfält"
                 :type-multiselect "Flervalsmeny"
                 :type-option "Rullgardinsmeny"
+                :type-phone-number "Phone number"
```
Comment (opqdonut) [inline resources/translations/sv.edn:269]: forgot to translate this? it's "telefonnummer" in swedish :)

File: resources/translations/sv.edn [hunk 4]
```diff
@@ -391,6 +392,7 @@
          :validation {:column-values-missing "Alla kolumner måste fyllas i i tabell \"%1\"."
                       :columns-required "Kräv minst en kolumn." ; TODO check
                       :invalid-email "Ogiltig e-postadress."
+                      :invalid-phone-number "Ogiltig phone number."
```
Comment (opqdonut) [inline resources/translations/sv.edn:395]: forgot to translate this?

File: src/clj/rems/api/schema.clj [hunk 5]
```diff
@@ -129,7 +129,7 @@
 ;;; template for a form field, before answering
 (s/defschema FieldTemplate
   {:field/id schema-base/FieldId
-   :field/type (s/enum :attachment :date :description :email :header :label :multiselect :option :text :texta :table)
+   :field/type (s/enum :attachment :date :description :email :header :label :multiselect :option :phonenumber :text :texta :table)
```
Comment (opqdonut) [inline src/clj/rems/api/schema.clj:132]: Minor nitpick: Would `:phone` be more in line with `:email`? If you want to call it a phone number, I think `:phone-number` would be more consistent with our api in general.

File: src/clj/rems/db/test_data.clj [hunk 6]
```diff
@@ -193,7 +193,14 @@
...
     :field/optional true
-    :field/max-length 100}])
+    :field/max-length 100}
+
+   {:field/type :phonenumber
+    :field/title {:en "Phone number"
+                  :fi "Puhelinnumero"
+                  :sv "Puhelinnumero"}
+    :field/optional true
+    :field/max-length 20}])
```
Comment (opqdonut) [inline src/clj/rems/db/test_data.clj:203]: if we have format checking for phone numbers, do we really need to also support max-length?

File: src/cljc/rems/common/form.cljc [hunk 7]
```diff
@@ -464,6 +464,19 @@
...
+    ;;                     [{:field/type :phonenumber
+    ;;                       :field/title {:en "en" :fi "fi"}
+    ;;                       :field/optional false}])]
+    ;;     (testing "valid form"
+    ;;       (is (empty? (validate-form-template form languages))))
+
+    ;;     (testing "missing title localization"
+    ;;       (is (= {:form/fields {0 {:field/title {:en :t.form.validation/required}}}}
+    ;;              (validate-form-template (assoc-in form [:form/fields 0 :field/title :en] "") languages)
+    ;;              (validate-form-template (assoc-in form [:form/fields 0 :field/title :en] nil) languages))))))
```
Comment (opqdonut) [inline src/cljc/rems/common/form.cljc:478]: A test like this would be nice. Remember to also test a valid non-empty value and an invalid non-empty value.

File: src/cljc/rems/common/util.cljc [hunk 8]
```diff
@@ -7,6 +7,8 @@
 ;; regex from https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/email#Validation
 (def +email-regex+ #"[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*")
 
+(def +phone-number-regex+ #"^\+\d{1,3}\s\d{2,3}\s\d{2,3}\s\d{4}|^\+\d{1,3}\s\d{1,14}(\s\d{1,13})?$|^\+[1-9]\d{5,14}$\s?$")
```
Comment (opqdonut) [inline src/cljc/rems/common/util.cljc:10]: Hmm, why not just `#"^\+[ 0-9]+`: a plus followed by spaces and numbers. We don't need to restrict how people enter spaces in their phone numbers, right?

Could also consider normalizing the phone number by removing spaces when it gets stored in the db (I think Jaakko requsted this), but that can be done in a follow-up PR.

File: src/cljs/rems/fields.cljs [hunk 9]
```diff
@@ -167,7 +167,7 @@
 (defn- event-value [event]
   (.. event -target -value))
 
-(defn text-field
+(defn phone-text-field
```
Comment (opqdonut) [inline src/cljs/rems/fields.cljs:170]: this looks like an unused function

File: test/clj/rems/test_pdf.clj [hunk 10]
```diff
@@ -162,7 +162,9 @@
                  [[:paragraph pdf/field-style "Text field with max length"]
                   [:paragraph "pdf test"]]
                  [[:paragraph pdf/field-style "Text area with max length"]
-                  [:paragraph "pdf test"]]]]]
+                  [:paragraph "pdf test"]]
+                 [[:paragraph {:spacing-before 8, :style :bold} "Phone number"]
```
Comment (opqdonut) [inline test/clj/rems/test_pdf.clj:166]: You should use `pdf/field-style` here as for the other cases. The idea is to avoid depending on the specific pdf styles and to only test the structure of the document.

**Review by opqdonut — CHANGES_REQUESTED** [review]
Looks good in general!

One additional thing that caught my eye: the PR title talks about IP addresses, but this doesn't implement those, right?

**Comment by marharyta** [comment]
@opqdonut I agree,  & IP address should be removed from the name, I was thinking of how to properly separate those features and arrived at a conclusion that separating phone number and IP address into two different PRs could be very helpful.

File: src/clj/rems/api/schema.clj (hunk 5 above)
Comment (marharyta) [reply src/clj/rems/api/schema.clj:132]: I can fix that, no big deal

File: src/clj/rems/db/test_data.clj (hunk 6 above)
Comment (marharyta) [reply src/clj/rems/db/test_data.clj:203]: I do not see the harm in it

Comment (jaakkocsc) [reply src/clj/rems/db/test_data.clj:203]: maximum length for a phone number is 15 digits, and then the plus sign (which actually is just a shorthand and gets replaced with 00 in the phone exchange).

File: src/cljc/rems/common/util.cljc (hunk 8 above)
Comment (jaakkocsc) [reply src/cljc/rems/common/util.cljc:10]: i agree with Joel, we should not care how people type in the number, but sanitise it when storing it in the db. this actually makes my previous comment about fixing the max length invalid, since we should allow people to type even "+ 1 2 3 4 5 6 7 8    9 10 11".

Comment (marharyta) [reply src/cljc/rems/common/util.cljc:10]: I could do that here!

File: src/cljs/rems/fields.cljs (hunk 9 above)
Comment (marharyta) [reply src/cljs/rems/fields.cljs:171]: Yes, I should have removed it while refactroing, my mistake

Comment (marharyta) [reply src/cljs/rems/fields.cljs:171]: I very often forget these things, I am too used to eslint throwing warnings about it that I cannot get from cloure tools unfortunately

File: src/clj/rems/db/test_data.clj (hunk 6 above)
Comment (opqdonut) [reply src/clj/rems/db/test_data.clj:203]: I just don't see a valid usecase for supporting max-length, and I think having max-length enabled for this field makes the form editor more complex.

**Comment by opqdonut** [comment]
Yeah, separate PRs for them sounds good!

File: test/clj/rems/test_pdf.clj (hunk 10 above)
Comment (marharyta) [reply test/clj/rems/test_pdf.clj]: addressed

File: src/cljs/rems/fields.cljs (hunk 9 above)
Comment (marharyta) [reply src/cljs/rems/fields.cljs]: addressed

File: src/clj/rems/db/test_data.clj (hunk 6 above)
Comment (marharyta) [reply src/clj/rems/db/test_data.clj:203]: I can fix that, should not be an issue

**Review by Chanakorn — APPROVED** [review]

File: src/cljs/rems/fields.cljs [hunk 11]
```diff
@@ -169,13 +169,18 @@
...
-        max-length (:field/max-length opts)]
+        max-length (:field/max-length opts)
+        type (:field/type opts)]
     [field-wrapper opts
-     [:input.form-control {:type "text"
+     [:input.form-control {:type (cond
+                                   (= type :text) "text"
+                                   (= type :phone-number) "tel"
+                                   (= type :email) "email"
+                                   :else "text")
```
Comment (Macroz) [inline src/cljs/rems/fields.cljs:183]: Use `case`? instead of this `cond`?

I think these are piggybacking a bit too much on the `text-field` function already. We could rename `text-field` to `basic-input-field` or something?

Comment (marharyta) [reply src/cljs/rems/fields.cljs:183]: I agree, case looks like a better fit here, also the text-field is misleading as we do not only put text there, Ill try to think of a better name.

Comment (marharyta) [reply src/cljs/rems/fields.cljs:183]: Also, @Macroz I am wondering how much date-field is just a repetition of text-field but with date?
Do you think it makes sense to unify them? Create one universal function for input field?

Comment (opqdonut) [reply src/cljs/rems/fields.cljs:183]: Seems to me like date-field could indeed be unified. Might be a good idea to do it in a separate PR though.

Comment (marharyta) [reply src/cljs/rems/fields.cljs:183]: ok, but good action point

Comment (marharyta) [reply src/cljs/rems/fields.cljs:183]: I ll create a small ticket

File: src/clj/rems/db/test_data.clj [hunk 12]
```diff
@@ -193,7 +193,13 @@
                   :fi "Tekstialue pituusrajalla"
                   :sv "Textområdet med begränsat längd"}
     :field/optional true
-    :field/max-length 100}])
+    :field/max-length 100}
+
+   {:field/type :phone-number
+    :field/title {:en "Phone number"
+                  :fi "Puhelinnumero"
+                  :sv "Puhelinnumero"}
```
Comment (Macroz) [inline src/clj/rems/db/test_data.clj:201]: Should be `Telefonnummer`?

Comment (marharyta) [reply src/clj/rems/db/test_data.clj:201]: True that

File: src/cljs/rems/fields.cljs (hunk 11 above)
Comment (Macroz) [reply src/cljs/rems/fields.cljs]: There doesn't need to be a new ticket btw. It can be a new bullet point to the current ticket since this is for it still.

File: src/clj/rems/form_validation.clj [hunk 13]
```diff
@@ -45,6 +45,14 @@
       {:field-id (:field/id field)
        :type     :t.form.validation/invalid-email})))
 
+(defn- invalid-phone-number-error [field]
+  (when (= (:field/type field) :phone-number)
+    (when-not (or (str/blank? (:field/value field))
+                  (re-matches +phone-number-regex+ (:field/value field)))
+      {:field-id (:field/id field)
+       :type     :t.form.validation/invalid-phone-number})))
```
[the remark this answers is not in the document]
Comment (Macroz) [reply src/clj/rems/form_validation.clj:53]: (it's ok to leave it like it is, just a note that maybe your editor settings are different)

[the remark this answers is not in the document]
Comment (opqdonut) [reply src/clj/rems/form_validation.clj:53]: There are some occurrences of this type of indentation, for example in rems.form-validation. I think they're leftovers from Leena or Esko who indented like this sometimes.

[the remark this answers is not in the document]
Comment (marharyta) [reply src/clj/rems/form_validation.clj:53]: lein cljfmt did not fix it, I assumed it is ok

[the remark this answers is not in the document]
Comment (opqdonut) [reply src/clj/rems/form_validation.clj:53]: Yeah, like Make said it's ok like this, just a FYI.

File: src/cljc/rems/common/util.cljc [hunk 14]
```diff
@@ -7,6 +7,17 @@
 ;; regex from https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/email#Validation
 (def +email-regex+ #"[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*")
 
+;; regex from https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/email#Validation
+(def +phone-number-regex+ #"^\+[1-9][0-9]{4,26}$")
```
Comment (marharyta) [inline src/cljc/rems/common/util.cljc:11]: works for  https://regexr.com/ but not for lein repl
need to investigate

**Comment by marharyta** [comment]
do not approve yet, I still have to fix the regex

**Comment by marharyta** [comment]
ready for re-review / merge

File: src/cljc/rems/common/util.cljc [hunk 15]
```diff
@@ -7,6 +7,19 @@
 ;; regex from https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/email#Validation
 (def +email-regex+ #"[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*")
 
+;; regex from https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/email#Validation
```
Comment (opqdonut) [inline src/cljc/rems/common/util.cljc:10]: is this comment true or just a copypaste error?

File: src/cljc/rems/common/util.cljc [hunk 16]
```diff
@@ -7,6 +7,19 @@
...
+(def +phone-number-regex+ #"^\+[1-9][0-9\s]{4,26}$")
```
Comment (opqdonut) [inline src/cljc/rems/common/util.cljc:11]: It's an ok regex but would it be better to allow 5-27 numerals and any number of spaces? Now e.g. `+1234567890123456789012345` is valid but `+123 456 789 012 345 678 901 2345` isn't.

File: src/cljs/rems/fields.cljs [hunk 17]
```diff
@@ -168,13 +168,18 @@
   (.. event -target -value))
 
 (defn text-field
```
Comment (opqdonut) [inline src/cljs/rems/fields.cljs:170]: were you going to figure out a better name for `text-field`?

File: src/clj/rems/form_validation.clj (hunk 13 above)
Comment (opqdonut) [inline src/clj/rems/form_validation.clj:53]: I'd like to see a test that exercises this code. Just a small new case in rems.test-form-validation is enough.

**Review by opqdonut — CHANGES_REQUESTED** [review]
Looks good. Please add the test and see my questions.

File: src/cljs/rems/fields.cljs (hunk 17 above)
Comment (opqdonut) [reply src/cljs/rems/fields.cljs:170]: If you were planning on doing it later along with the date-field refactor that was discussed earlier, that's fine too!

File: src/cljc/rems/common/util.cljc (hunk 15 above)
Comment (marharyta) [reply src/cljc/rems/common/util.cljc:10]: copypaster error

File: src/cljc/rems/common/util.cljc (hunk 16 above)
Comment (marharyta) [reply src/cljc/rems/common/util.cljc:11]: true, but I did not want to make it too complex

File: src/cljs/rems/fields.cljs (hunk 17 above)
Comment (marharyta) [reply src/cljs/rems/fields.cljs:170]: i created a separate issue for refactoring this one

File: src/clj/rems/form_validation.clj (hunk 13 above)
Comment (marharyta) [reply src/clj/rems/form_validation.clj:53]: I can do that np

File: src/cljs/rems/fields.cljs (hunk 17 above)
Comment (opqdonut) [reply src/cljs/rems/fields.cljs:170]: looks like it's #2600, added a note comment about renaming `text-field` as well

File: src/cljc/rems/common/util.cljc (hunk 16 above)
Comment (marharyta) [reply src/cljc/rems/common/util.cljc:10]: i think it can be like this for now, we can alter it later if we see fit

Comment (opqdonut) [reply src/cljc/rems/common/util.cljc:10]: fine by me

**Review by opqdonut — APPROVED** [review]
feel free to merge after addressing my comment

File: test/clj/rems/test_form_validation.clj [hunk 18]
```diff
@@ -356,6 +356,37 @@
...
+             (validate-fields-for-draft [{:field/id       "1"
+                                          :field/type     :phone-number
+                                          :field/optional false
+                                          :field/visible  true
+                                          :field/value    ""}
+                                         {:field/id       "2"
+                                          :field/type     :phone-number
+                                          :field/optional false
+                                          :field/visible  true
+                                          :field/value    "+058450000100"}
```
[the remark this answers is not in the document]
Comment (marharyta) [reply test/clj/rems/test_form_validation.clj:374]: sure!

File: test/clj/rems/test_form_validation.clj [hunk 19]
```diff
@@ -356,6 +356,37 @@
...
+                                         {:field/id       "3"
+                                          :field/type     :phone-number
+                                          :field/optional false
+                                          :field/visible  true
+                                          :field/value    "+050050000100000000000000"}
```
Comment (Macroz) [inline test/clj/rems/test_form_validation.clj:379]: Is this one too long? doesn't it fail also because of the zero prefix?

File: test/clj/rems/test_form_validation.clj [hunk 20]
```diff
@@ -356,6 +356,37 @@
...
+      (is (= [{:type :t.form.validation/invalid-phone-number :field-id "2"}
+              {:type :t.form.validation/invalid-phone-number :field-id "3"}
+              {:type :t.form.validation/invalid-phone-number :field-id "4"}
+              {:type :t.form.validation/invalid-phone-number :field-id "5"}]
+             (validate-fields-for-draft [{:field/id       "1"
+                                          :field/type     :phone-number
+                                          :field/optional false
+                                          :field/visible  true
+                                          :field/value    ""}
+                                         {:field/id       "2"
```
Comment (Macroz) [inline test/clj/rems/test_form_validation.clj:370]: The `:field/id` could be used to document the test case. It's always good to try to show the purpose of each case.

Comment (Macroz) [reply test/clj/rems/test_form_validation.clj:370]: This could be so `:field/id "prefix 0 is not allowed"`

Comment (marharyta) [reply test/clj/rems/test_form_validation.clj:370]: sounds like a really good idea!

File: test/clj/rems/test_form_validation.clj (hunk 19 above)
Comment (marharyta) [reply test/clj/rems/test_form_validation.clj:379]: true, fixed that in the tests

File: test/clj/rems/test_form_validation.clj (hunk 18 above)
[the remark this answers is not in the document]
Comment (marharyta) [reply test/clj/rems/test_form_validation.clj:374]: done

File: test/clj/rems/test_form_validation.clj [hunk 21]
```diff
@@ -357,6 +357,42 @@
                                         :field/visible  true
                                         :field/value    "invalid.email"}]))))
 
+  (testing "error: field input invalid phone number"
+    (is (= [{:type :t.form.validation/invalid-phone-number :field-id "less than 5 symbols is not allowed"}
+            {:type :t.form.validation/invalid-phone-number :field-id "more than 25 symbold is not allowed"}
+            {:type :t.form.validation/invalid-phone-number :field-id "numer has to start with a + prefix"}
```
Comment (opqdonut) [inline test/clj/rems/test_form_validation.clj:363]: typo: numer

File: test/clj/rems/test_form_validation.clj [hunk 22]
```diff
@@ -357,6 +357,42 @@
...
+                                        :field/type     :phone-number
+                                        :field/optional false
+                                        :field/visible  true
+                                        :field/value    ""}
+                                       {:field/id       "prefix 0 is not allowed"
+                                        :field/type     :phone-number
+                                        :field/optional false
+                                        :field/visible  true
+                                        :field/value    "+058450000100"}
+                                       {:field/id       "more than 25 symbold is not allowed"
```
Comment (opqdonut) [inline test/clj/rems/test_form_validation.clj:375]: typo: symbold

**Review by opqdonut — APPROVED** [review]
feel free to merge

noticed a couple of minor typos

**Comment by marharyta** [comment]
merging after tests pass
